Standards/scs-0101-v1-entropy.md

```diff
@@ -5,114 +5,118 @@ status: Draft
+the HW RNG must be fed into the pool (if so desired) via the daemon `rngd`.
+The kernel converts the entropy from the pool into cryptographically
+secure random numbers that appear under `/dev/random`.
+
+With kernel 5.18, the algorithm that accomplishes
+said conversion has been drastically improved (see
+[linux-rng-5.17-18](https://web.archive.org/web/20230321040526/https://www.zx2c4.com/projects/linux-rng-5.17-5.18/)),
+so much so that running out of entropy is virtually ruled out.
+These patches have now also arrived in the upstream LTS images.
+
+### Entropy in virtual instances
+
+Virtual instances or virtual machines do not have the traditional sources
+of entropy mentioned above. However, the more recent methods mentioned
+above do work just fine (the RDRAND instruction set is not privileged).
+
+Alternatively, a virtualized HW RNG called `virtio-rng` can be established
+that injects entropy from the host into the instance, where this
+entropy can be sourced optionally from either the host's `/dev/random` or
+some HW RNG in the host. This virtualized HW RNG behaves just like real
+one, that is, it appears as `/dev/hwrng`, and the daemon `rngd` must
+be used to feed it into the kernel's entropy pool.
+
+In summary, with current kernels and CPUs entropy in virtual instances
+is readily available to a sufficient degree. In addition, the host's
+entropy sources can be injected using `virtio-rng` if so desired, e.g.,
+to enable access to a HW RNG.
+
+## Motivation
+
+As stated above, good sources of entropy are paramount for many
+important applications. Moreover, current technology makes it easy
+to provide these sources to virtual instances. Therefore, this standard
+mandates that these sources be made available on all conformant clouds.
+
+## Entropy in SCS clouds
+
+This standard does not intend to make any guarantees for linux images
+that are not themselves included in the
+[SCS standard](https://github.com/SovereignCloudStack/standards/blob/main/Standards/scs-0102-v1-image-metadata.md), in particular
+images having a kernel below version 5.18.
+
+### Flavors
+
+All flavors must have the following attribute activated:
 
 ```console
 hw_rng:allowed=True
 ```
 
-optional:
+The following attributes are optional:
 
 ```console
 hw_rng:rate_bytes - The allowed amount of bytes for the the guest
-to read from the host’s entropy per period.
-
-hw_rng:rate_period
+    to read from the host's entropy per period.
+hw_rng:rate_period - Sets the duration of a read period in seconds.
 ```
 
-#### 1.3.2 Images
+### Images
 
 Images must activate the attribute `hw_rng_model: virtio`.
 
-#### 1.3.3 Compute Nodes
+The daemon `rngd` must be installed (usually from `rng-tools`
+or `rng-utils`).
+
+The user may choose to use the `virtio-rng` device via `rngd`.
+
+### Compute nodes
+
+Compute nodes must use CPUs that offer RDRAND/RDSEED. This requirement
+is both very hard to verify and almost impossible to violate, so it will
+not be tested for.
```

Comment:
Why must RDRAND/RDSEED be necessary? I couldn't find a hard requirement for them in the linux kernel or anywhere else (e.g. OpenSSL also only lists them as an optional tool to generate their keys and so on).
https://en.wikipedia.org/wiki/RDRAND and the connected sources state that it was a source for randomness in previous kernel version, but not the only one in the pool, and since the algorithm was changed in favor of using CPU jitter (which doesn't come from RDRAND/RDSEED), this requirement should not be necessary in my opinion.

Reply:
My interpretation of the Donenfeld writeup was that his algorithm uses the first instruction of the following list that is supported by the CPU: RDSEED, RDRAND, RDTSC. It suggested to me that RDSEED was somehow superior.

Reply:
Yes it uses these Operations in this order, but it doesn't decrease the value of the entropy significantly if one of the later methods is used, since it is only mixed in as an entropy source, see

> I touched on RDSEED usage earlier, and there are in fact some more detailed changes on that topic. In [random: use RDSEED instead of RDRAND in entropy extraction](https://web.archive.org/web/20230321040526/https://git.kernel.org/crng/random/c/28f425e573e9), the RNG now uses RDSEED, if available, before falling back to RDRAND and eventually RDTSC. Since the CPU’s RNG is just another entropy input source — one we happen to call at boot time for initial seeding and during each subsequent reseeding — we might as well use its best source. We’re not using it as part of the direct generation of random numbers itself, but just as an additional entropy input, and hence there’s not a need for the speed improvements of RDRAND over RDSEED. (Incidentally I recently [removed direct RDRAND-usage from systemd](https://web.archive.org/web/20230321040526/https://github.com/systemd/systemd/commit/ffa047a03e4c5f6bd3af73b7eecb99cd230fe204) too.)

So I wouldn't write this in as a requirement, especially since most modern x86-CPUs would fullfill these requirements anyway.

Reply:
I agree with you that mandating this is kind of redundant (because the instructions are present anyway, and because the kernel works fine without them).

Reply:
I will try to get more feedback in tomorrow's meeting, and if nobody intervenes, I will remove this requirement.